core/controllers/base_test.py

```diff
@@ -209,6 +229,52 @@ def test_root_redirect_rules_for_logged_in_editors(self):
         self.assertIn('dashboard', response.headers['location'])
         self.logout()
 
+    def test_get_with_invalid_return_type_logs_correct_warning(self):
+        # Modify the testapp to use the mock handler.
+        self.testapp = webtest.TestApp(webapp2.WSGIApplication(
+            [webapp2.Route('/mock', self.MockHandler, name='MockHandler')],
+            debug=feconf.DEBUG,
+        ))
+
+        observed_log_messages = []
+        def mock_logging_function(msg, *_):
+            observed_log_messages.append(msg)
+
+        with self.swap(logging, 'warning', mock_logging_function):
+            self.get_json('/mock', expected_status_int=500)
+            self.assertEqual(len(observed_log_messages), 1)
+            self.assertEqual(
+                observed_log_messages[0],
+                'Not a recognized return type: defaulting to render JSON.')
+
+    def test_unrecognized_request_method_logs_correct_warning(self):
+        self.testapp = webtest.TestApp(webapp2.WSGIApplication(
+            [webapp2.Route('/mock', self.MockHandler, name='MockHandler')],
+            debug=feconf.DEBUG,
+        ))
+
+        observed_log_messages = []
+        def mock_logging_function(msg, *_):
+            observed_log_messages.append(msg)
+
+        with self.swap(logging, 'warning', mock_logging_function):
+            self.testapp.head('/mock', status=500)
+            self.assertEqual(len(observed_log_messages), 2)
+            self.assertEqual(
+                observed_log_messages[0],
+                'Not a recognized request method.')
+            self.assertEqual(
+                observed_log_messages[1],
+                'Not a recognized return type: defaulting to render JSON.')
+
+    def test_renders_error_page_with_iframed(self):
+        # Modify the testapp to use the mock handler.
+        self.testapp = webtest.TestApp(webapp2.WSGIApplication(
+            [webapp2.Route('/mock1', self.MockHandler1, name='MockHandler1')],
+            debug=feconf.DEBUG,
+        ))
+        self.get_html_response('/mock1', expected_status_int=500)
```

Comment:
It renders a HTML error template. I did an assertIn to match the contents of the page with the reponse.body. It returns a 500 error code as we are rendering an invalid template

Reply:
Sorry, not sure I understand. If there's no template to render then how exactly are we showing "Uh-oh! The Oppia exploration ... deleted."?

Reply:
We are first getting an error that the template is invalid. And that error together with the 'True' value of iframed renders the iframe_error page. This happens from within base.py

Reply:
I see, OK. Add a comment here to say that the 500 is expected because the template file does not exist (so it is a legitimate server error caused by our mock).

In general every 500 should be explained, since 500s are not expected.

Reply:
Done